astropy/stats/_fast_sigma_clipping.c

```diff
@@ -0,0 +1,239 @@
+
+double kth_smallest(double a[], int n, int k)
+{
+    register int i,j,l,m ;
+    register double x ;
+
+    l=0 ; m=n-1 ;
+    while (l<m) {
+        x=a[k] ;
+        i=l ;
+        j=m ;
+        do {
+            while (a[i]<x) i++ ;
+            while (x<a[j]) j-- ;
+            if (i<=j) {
+                ELEM_SWAP(a[i],a[j]) ;
+                i++ ; j-- ;
+            }
+        } while (i<=j) ;
+        if (j<k) l=i ;
+        if (k<i) m=j ;
+    }
+    return a[k] ;
+}
+
+double wirth_median(double a[], int n) {
+  if (n % 2 == 0) {
+    return 0.5 * (kth_smallest(a,n,n/2) + kth_smallest(a,n,n/2 - 1));
+  } else {
+    return kth_smallest(a,n,(n-1)/2);
+  }
+}
+
+static PyObject *_sigma_clip_fast(PyObject *self, PyObject *args) {
+
+  long n, m;
+  int i, j;
+  double *buffer;
+  PyObject *data_obj, *mask_obj;
+  PyArrayObject *data_array, *mask_array;
+  double *data;
+  uint8_t *mask;
+  int use_median, maxiters;
+  double sigma_lower, sigma_upper, mean, median, std;
+  int iteration, count;
+  double lower, upper;
+  int new_count;
+
+  // Parse the input tuple
+  if (!PyArg_ParseTuple(args, "OOiidd", &data_obj, &mask_obj, &use_median,
+                        &maxiters, &sigma_lower, &sigma_upper)) {
+    PyErr_SetString(PyExc_TypeError, "Error parsing input");
+    return NULL;
+  }
+
+  // Interpret the input data array as a `numpy` array.
+  data_array = (PyArrayObject *)PyArray_FROM_O(data_obj);
+  if (data_array == NULL) {
+    PyErr_SetString(PyExc_TypeError, "Couldn't parse the input data array.");
+    Py_XDECREF(data_array);
+    return NULL;
+  }
+
+  // Interpret the input mask array as a `numpy` array.
+  mask_array = (PyArrayObject *)PyArray_FROM_O(mask_obj);
+  if (mask_array == NULL) {
+    PyErr_SetString(PyExc_TypeError, "Couldn't parse the input mask array.");
+    Py_XDECREF(mask_array);
+    return NULL;
+  }
+
+  // Get the data array shape
+  n = (long)PyArray_DIM(data_array, 0);
+  m = (long)PyArray_DIM(data_array, 1);
+
+  // Build the temporary array and mask
+  buffer = (double *)malloc(n * sizeof(double));
+  if (buffer == NULL) {
+    PyErr_SetString(PyExc_RuntimeError, "Couldn't build buffer array");
+    Py_DECREF(data_array);
+    Py_DECREF(mask_array);
+    return NULL;
+  }
+
+  // At this point we should ideally use the iterator API in Numpy,
+  // but for simplicity for now we assume the data is contiguous in
+  // memory and will deal with the correct iteration later
+
+  data = (double *)PyArray_DATA(data_array);
+  mask = (uint8_t *)PyArray_DATA(mask_array);
+
+  // This function is constructed to take a 2-d array of values and assumes
+  // that each 1-d array when looping over the last dimension should be
+  // treated separately.
+
+  for (j = 0; j < m; j++) {
+
+    iteration = 0;
+
+    // We copy all finite values from array into the buffer
+
+    count = 0;
+    for (i = 0; i < n; i++) {
+      if (data[i * m + j] == data[i * m + j]) {
```

Comment:
My sense would be that rather than rely on non-finite elements not comparing equal, it might make sense to instead start with an input mask. But that's really an implementation detail.

Reply:
Done